src/store/multi.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use lmdb;
+
+use lmdb::{
+    Cursor,
+    Database,
+    Iter as LmdbIter,
+    //    IterDup as LmdbIterDup,
+    RoCursor,
+    RwTransaction,
+    Transaction,
+    WriteFlags,
+};
+
+use crate::error::StoreError;
+
+use crate::value::{
+    OwnedValue,
+    Value,
+};
+
+fn read_transform(val: Result<&[u8], lmdb::Error>) -> Result<Option<Value>, StoreError> {
+    match val {
+        Ok(bytes) => Value::from_tagged_slice(bytes).map(Some).map_err(StoreError::DataError),
+        Err(lmdb::Error::NotFound) => Ok(None),
+        Err(e) => Err(StoreError::LmdbError(e)),
+    }
+}
+
+fn read_transform_owned(val: Result<&[u8], lmdb::Error>) -> Result<Option<OwnedValue>, StoreError> {
+    match val {
+        Ok(bytes) => Value::from_tagged_slice(bytes).map(|v| Some(OwnedValue::from(&v))).map_err(StoreError::DataError),
+        Err(lmdb::Error::NotFound) => Ok(None),
+        Err(e) => Err(StoreError::LmdbError(e)),
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct MultiStore {
+    db: Database,
+}
+
+pub struct Iter<'env> {
+    iter: LmdbIter<'env>,
+    cursor: RoCursor<'env>,
+}
+
+impl MultiStore {
+    pub(crate) fn new(db: Database) -> MultiStore {
+        MultiStore {
+            db,
+        }
+    }
+
+    /// Provides a cursor to all of the values for the duplicate entries that match this key
+    pub fn get<'env, T: Transaction, K: AsRef<[u8]>>(&self, txn: &'env T, k: K) -> Result<Iter<'env>, StoreError> {
+        let mut cursor = txn.open_ro_cursor(self.db).map_err(StoreError::LmdbError)?;
+        let iter = cursor.iter_dup_of(k);
+        Ok(Iter {
+            iter,
+            cursor,
+        })
+    }
+
+    /// Provides a cursor to all of the values for the duplicate entries that match this key
+    pub fn get_first<'env, T: Transaction, K: AsRef<[u8]>>(
+        &self,
+        txn: &'env T,
+        k: K,
+    ) -> Result<Option<Value<'env>>, StoreError> {
+        let result = txn.get(self.db, &k);
+        read_transform(result)
+    }
+
+    /// Insert a value at the specified key.
+    /// This put will allow duplicate entries.  If you wish to have duplicate entries
+    /// rejected, use the `put_with_flags` function and specify NO_DUP_DATA
```

Comment:
Nit: we should add tests for the _put_with_flags()_ functions as well, although we can do that in a followup.